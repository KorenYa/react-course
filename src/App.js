import React, { Component } from 'react';
import ErrorBoundary from './components/ErrorBoundary';
import SearchForm from './components/SearchForm';
import StatusBar from './components/StatusBar';
import SearchResult from './components/SearchResult';

import './styles/general.scss';

class App extends Component {
    constructor(props) {
        super(props);
        this.state = { inputValue: 'Narnia', moviesList: { total: 0 } };

        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);
    }

    handleChange(event) {
        this.setState({ inputValue: event.target.value });
    }

    handleSubmit(event) {
        event.preventDefault();
        fetch(
            'http://react-cdp-api.herokuapp.com/movies?search=' +
                this.state.inputValue +
                '&searchBy=title'
        )
            .then(response => response.json())
            .then(moviesList => this.setState({ moviesList }));
    }

    render() {
        return (
            <ErrorBoundary>
                <SearchForm
                    inputValue={this.state.inputValue}
                    handleChange={this.handleChange}
                    handleSubmit={this.handleSubmit}
                />
                <StatusBar moviesList={this.state.moviesList} />
                <SearchResult moviesList={this.state.moviesList} />
            </ErrorBoundary>
        );
    }
}
export default App;
