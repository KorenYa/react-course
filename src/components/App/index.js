import React, { Component } from 'react';
import ErrorBoundary from '../ErrorBoundary';
import SearchForm from '../SearchForm';
import StatusBar from '../StatusBar';
import SearchResult from '../SearchResult';
import MovieDetails from '../MovieDetails';
import { connect } from 'react-redux';
import {
    startMoviesSearch,
    updateSearchQuery,
    updateSearchBy,
    updateSortBy
} from '../../store/AC';

import '../../styles/general.scss';

export class App extends Component {
    constructor(props) {
        super(props);

        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.handleFilter = this.handleFilter.bind(this);
        this.handleSortBy = this.handleSortBy.bind(this);
    }

    handleSubmit(event) {
        event.preventDefault();
        this.props.startMoviesSearch();
    }

    handleChange(event) {
        this.props.updateSearchQuery(event.target.value);
    }

    handleFilter(event, searchBy) {
        event.preventDefault();
        this.props.updateSearchBy(searchBy);
    }

    handleSortBy(sortBy) {
        this.props.updateSortBy(sortBy);
        this.props.startMoviesSearch();
    }

    render() {
        const { searchQuery, searchBy, moviesList, sortBy } = this.props;

        return (
            <ErrorBoundary>
                <SearchForm
                    inputValue={searchQuery}
                    handleChange={this.handleChange}
                    handleSubmit={this.handleSubmit}
                    searchBy={searchBy}
                    handleFilter={this.handleFilter}
                />
                <StatusBar
                    total={moviesList.total}
                    sortBy={sortBy}
                    handleSortBy={this.handleSortBy}
                />
                <SearchResult moviesList={moviesList} />
                {moviesList.total > 0 && (
                    <MovieDetails movie={moviesList.data[0]} />
                ) /*will be moved in separate route*/}
            </ErrorBoundary>
        );
    }
}
export default connect(
    state => {
        return {
            moviesList: state.moviesList,
            searchQuery: state.searchQuery,
            searchBy: state.searchBy,
            sortBy: state.sortBy
        };
    },
    { startMoviesSearch, updateSearchQuery, updateSearchBy, updateSortBy }
)(App);
