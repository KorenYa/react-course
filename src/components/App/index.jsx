import React, { Component } from 'react';
import { withRouter, Route, Switch } from 'react-router-dom';
import { connect } from 'react-redux';
import ErrorBoundary from '../ErrorBoundary';

import SearchForm from '../SearchForm';
import StatusBar from '../StatusBar';
import SearchResult from '../SearchResult';
import MovieDetails from '../MovieDetails';
import NotFound from '../NotFound';

import {
  startMoviesSearch,
  updateSearchQuery,
  updateSearchBy,
  updateSortBy,
  clearMoviesList,
} from '../../store/AC';

import '../../styles/general.scss';

export class App extends Component {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleFilter = this.handleFilter.bind(this);
    this.handleSortBy = this.handleSortBy.bind(this);
  }

  componentDidMount() {
    const {
      searchQuery,
      updateSearchQuery,
      startMoviesSearch,
      clearMoviesList,
    } = this.props;
    const { location: pathname } = this.props;
    const browserSearchQuery = '';

    // Scenarios how to handle persisted store results and first load based on path
    // 1) if user opens homepage '/' => purge store and show 'No results' page
    if (pathname && pathname === '/') {
      updateSearchQuery('');
      clearMoviesList();
    }

    // 2) if user opens search page'/search/{browserSearchQuery}' with new request
    // => override store with new response
    if (browserSearchQuery && browserSearchQuery !== searchQuery) {
      updateSearchQuery(browserSearchQuery);
      startMoviesSearch();
    }

    // 3) if user opens search page'/search/{browserSearchQuery}' with already handled request
    // => show response from store
    // auto resolved
  }

  handleSubmit(event) {
    event.preventDefault();
    const { history, searchQuery, startMoviesSearch } = this.props;
    startMoviesSearch();
    history.push(`/search/${searchQuery}`);
  }

  handleChange(event) {
    const { updateSearchQuery } = this.props;
    updateSearchQuery(event.target.value);
  }

  handleFilter(event, searchBy) {
    event.preventDefault();
    const { updateSearchBy } = this.props;
    updateSearchBy(searchBy);
  }

  handleSortBy(sortBy) {
    const { updateSortBy, startMoviesSearch } = this.props;
    updateSortBy(sortBy);
    startMoviesSearch();
  }

  render() {
    const {
      searchQuery, searchBy, moviesList, sortBy,
    } = this.props;

    const Homepage = () => (
      <React.Fragment>
        <SearchForm
          inputValue={searchQuery}
          handleChange={this.handleChange}
          handleSubmit={this.handleSubmit}
          searchBy={searchBy}
          handleFilter={this.handleFilter}
        />
        <StatusBar
          total={moviesList.total}
          sortBy={sortBy}
          handleSortBy={this.handleSortBy}
        />
        <SearchResult moviesList={moviesList} />
      </React.Fragment>
    );

    return (
      <ErrorBoundary>
        <Switch>
          <Route exact path="/" render={Homepage} />
          <Route
            path="/search/:browserSearchQuery"
            render={Homepage}
          />
          <Route
            path="/film/:filmId"
            render={props => <MovieDetails {...props} />}
          />
          <Route path="*" component={NotFound} />
        </Switch>
      </ErrorBoundary>
    );
  }
}

export default withRouter(
  connect(
    state => ({
      moviesList: state.moviesList,
      searchQuery: state.searchQuery,
      searchBy: state.searchBy,
      sortBy: state.sortBy,
    }),
    {
      startMoviesSearch,
      updateSearchQuery,
      updateSearchBy,
      updateSortBy,
      clearMoviesList,
    },
  )(App),
);
