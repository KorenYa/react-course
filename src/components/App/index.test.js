import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { App } from '../App';

configure({ adapter: new Adapter() });

describe('App', () => {
    const location = {
        pathname: '*'
    };

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <MemoryRouter>
                    <App
                        moviesList={{ total: 0 }}
                        searchQuery={'Adventure'}
                        searchBy={'title'}
                        sortBy={'release_date'}
                        location={location}
                    />
                </MemoryRouter>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    const mockStartMoviesSearch = jest.fn();
    const mockUpdateSearchQuery = jest.fn();
    const mockUpdateSearchBy = jest.fn();
    const mockUpdateSortBy = jest.fn();
    const mockClearMoviesList = jest.fn();
    const event = {
        preventDefault: jest.fn(),
        target: {
            value: 'mock'
        }
    };

    const history = {
        push: jest.fn()
    };
    const wrapper = shallow(
        <App
            moviesList={{ total: 0 }}
            searchQuery={'Adventure'}
            searchBy={'title'}
            sortBy={'release_date'}
            startMoviesSearch={mockStartMoviesSearch}
            updateSearchQuery={mockUpdateSearchQuery}
            updateSearchBy={mockUpdateSearchBy}
            updateSortBy={mockUpdateSortBy}
            clearMoviesList={mockClearMoviesList}
            location={location}
            history={history}
        />
    );
    const instance = wrapper.instance();

    it('handleSubmit: should call mockStartMoviesSearch function', () => {
        instance.handleSubmit(event);
        expect(mockStartMoviesSearch.mock.calls.length).toBe(1);
    });

    it('handleChange: should call mockUpdateSearchQuery function', () => {
        instance.handleChange(event);
        expect(mockUpdateSearchQuery.mock.calls.length).toBe(1);
    });

    it('handleFilter: should call mockUpdateSearchBy function', () => {
        instance.handleFilter(event, 'mock');
        expect(mockUpdateSearchBy.mock.calls.length).toBe(1);
    });

    it('handleSortBy: should call mockUpdateSortBy && mockStartMoviesSearch functions', () => {
        instance.handleSortBy(event, 'mock');
        expect(mockUpdateSortBy.mock.calls.length).toBe(1);
        expect(mockStartMoviesSearch.mock.calls.length).toBe(2);
    });
});
