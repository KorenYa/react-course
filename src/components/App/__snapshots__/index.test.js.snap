// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
Array [
  <form
    className="search-form"
    onSubmit={[Function]}
  >
    <div
      className="search-form__title"
    >
      netflixroulette
    </div>
    <h1
      className="title"
    >
      find your movie
    </h1>
    <input
      className="input"
      onChange={[Function]}
      type="text"
      value="Adventure"
    />
    <div
      className="search-form__btn-wrapper"
    >
      <span
        className="search-form__subtitle"
      >
        Search by
      </span>
      <button
        className="btn active"
        data-searchby="title"
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
      <button
        className="btn "
        data-searchby="genre"
        onClick={[Function]}
        type="button"
      >
        Genre
      </button>
      <button
        className="btn btn-lg"
        onClick={[Function]}
        type="submit"
      >
        Search
      </button>
    </div>
  </form>,
  <div
    className="status-bar"
  >
    <span>
      Â 
    </span>
  </div>,
  <div
    className="search-result__no-hits"
  >
    No films found
  </div>,
]
`;
